src/charm.py

```diff
@@ -498,14 +502,195 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
         logger.info("Backup uploaded to S3 bucket %s", s3_parameters["bucket"])
         event.set_results({"backup-id": backup_key})
 
+    def _on_list_backups_action(self, event: ActionEvent) -> None:
+        """Handles list-backups action.
+
+        Lists all backups stored in S3 bucket.
+
+        Args:
+            event: ActionEvent
+        """
+        if not self._is_relation_created(S3_RELATION_NAME):
+            logger.error("S3 relation not created. Failed to list backups.")
+            event.fail(message="S3 relation not created. Failed to list backups.")
+            return
+
+        if not self.unit.is_leader():
+            logger.error("Only leader unit can list backups.")
+            event.fail(message="Only leader unit can list backups.")
+            return
+
+        missing_parameters = self._get_missing_s3_parameters()
+        if missing_parameters:
+            logger.error("S3 parameters missing. %s", missing_parameters)
+            event.fail(message=f"S3 parameters missing. {missing_parameters}")
+            return
+        s3_parameters = self._retrieve_s3_parameters()
+
+        try:
+            s3 = S3(
+                access_key=s3_parameters["access-key"],
+                secret_key=s3_parameters["secret-key"],
+                endpoint=s3_parameters["endpoint"],
+                region=s3_parameters.get("region"),
+            )
+        except (BotoCoreError, ClientError, ValueError) as e:
+            logger.error("Failed to create S3 session: %s", e)
+            event.fail(message="Failed to create S3 session.")
+            return
```

Comment:
This exact code is repeated 3 times now, once for each action. Would it make sense to extract it to a separate method? It could make it simpler to maintain in the future, for things like changing log messages or renaming parameters. WDYT?

Reply:
I'm hesitant between the following approaches.
1. Having a function that we pass the event to as a parameter and it does the requirement check
My problem with this approach is that this function will be coupled to the context of the action handlers.
```
def _check_backup_restore_actions_s3_requirements(self, event: ActionEvent) -> None:
        """Checks the conditions for creating S3 in backup and restore actions.

        Args:
            event: ActionEvent

        Returns:
            bool: True if the conditions are met, False otherwise.
        """
        if not self._is_relation_created(S3_RELATION_NAME):
            logger.error("S3 relation not created. Failed to run backup/restore operation.")
            event.fail(message="S3 relation not created. Failed to run backup/restore operation.")
            return

        if not self.unit.is_leader():
            logger.error("Only leader unit can run backup/restore operation.")
            event.fail(message="Only leader unit can run backup/restore operation.")
            return

        missing_parameters = self._get_missing_s3_parameters()
        if missing_parameters:
            logger.error("S3 parameters missing. %s", missing_parameters)
            event.fail(message=f"S3 parameters missing. {missing_parameters}")
```

2. Having a more specific requirements check that will not validate everything, but will be scoped to the requirements of S3 (similar one for Vault requirements)
This is better as it's not coupled and can be used in different contexts if needed, the problem with is that it will need to return a message and some of the checks will be kept in the action handler causing a few cases of code duplication (for example checking leader)

```
def _check_s3_requirements(self) -> Tuple[bool, Optional[str]]:
        """Validates the conditions for creating S3.

        It will check if the S3 relation is created
            and if the required S3 parameters are set.

        Returns:
            bool: True if the conditions are met, False otherwise.
        """
        if not self._is_relation_created(S3_RELATION_NAME):
            return False, "S3 relation not created."

        missing_parameters = self._get_missing_s3_parameters()
        if missing_parameters:
            return False, f"S3 parameters missing. {missing_parameters}"

        return True, None
```

I feel there is a trade-off here, and personally I prefer not to couple functions to each other, therefore, I prefer the second choice and I have pushed a commit using this solution (wiht the Vault requirements checks too)

Let me know what do you think and if you see a different approach or if we can do something else to improve this.

Reply:
This works for me.